crates/router/src/core/payments/operations/payment_confirm.rs

```diff
@@ -718,6 +756,174 @@ impl<F: Clone, Ctx: PaymentMethodRetrieve>
+                fingerprint_hash.map(|id| consts::BASE64_ENGINE.encode(id));
+            let mut fingerprint_id = None;
+
+            //validating the payment method.
+            let mut is_pm_blocklisted = false;
+            let find_for_fingerprint = db
+                .find_blocklist_lookup_entry_by_merchant_id_kms_decrypted_hash(
+                    merchant_id.clone(),
+                    fingerprint_encoded_hash
+                        .clone()
+                        .ok_or(errors::ApiErrorResponse::InternalServerError)?,
+                );
+
+            let find_for_cardbin = db
+                .find_blocklist_lookup_entry_by_merchant_id_kms_decrypted_hash(
+                    merchant_id.clone(),
+                    cardbin_encoded_hash
+                        .clone()
+                        .ok_or(errors::ApiErrorResponse::InternalServerError)?,
+                );
+
+            let find_for_extended_cardbin = db
+                .find_blocklist_lookup_entry_by_merchant_id_kms_decrypted_hash(
+                    merchant_id.clone(),
+                    extended_cardbin_encoded_hash
+                        .clone()
+                        .ok_or(errors::ApiErrorResponse::InternalServerError)?,
+                );
+
+            let (is_fingerprint_blocked, is_cardbin_blocked, is_extended_cardbin_blocked) = tokio::join!(
+                find_for_fingerprint,
+                find_for_cardbin,
+                find_for_extended_cardbin
+            );
+
+            match (
+                is_fingerprint_blocked,
+                is_cardbin_blocked,
+                is_extended_cardbin_blocked,
+            ) {
+                (Ok(_), _, _) | (_, Ok(_), _) | (_, _, Ok(_)) => {
+                    intent_status = storage_enums::IntentStatus::Failed;
+                    is_pm_blocklisted = true;
+                }
+                _ => {
+                    let kms_hash = kms::get_kms_client(&state.conf.kms)
+                        .await
+                        .encrypt(
+                            fingerprint_encoded_hash
+                                .ok_or(errors::ApiErrorResponse::InternalServerError)?
+                                .as_bytes(),
+                        )
+                        .await
+                        .change_context(errors::ApiErrorResponse::InternalServerError)?;
+
+                    fingerprint_id = Some(
+                        db.insert_pm_fingerprint_entry(
+                            diesel_models::pm_fingerprint::PmFingerprintNew {
+                                fingerprint_id: utils::generate_id(
+                                    consts::ID_LENGTH,
+                                    "fingerprint",
+                                ),
+                                kms_hash,
+                            },
+                        )
+                        .await
+                        .change_context(errors::ApiErrorResponse::InternalServerError)?
```

Comment:
We probably don't want to throw an error here and fail the payment just because we failed to save the fingerprint to DB. Just log the error and move on. Same for KMS encryption above.